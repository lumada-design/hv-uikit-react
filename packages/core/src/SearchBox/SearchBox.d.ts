import { StandardProps } from "@material-ui/core";
import { HvInputProps } from "../Input";

export type HvSearchBoxClassKey = "root";

export interface HvSearchBoxProps
  extends StandardProps<HvInputProps, HvSearchBoxClassKey, "onSubmit"> {
  /**
   * The function that will be executed on Enter, allows checking the value state,
   * it receives the value or event+value.
   */
  onSubmit?: (event: Event, value: string) => void;
}

export default function HvSearchBox(props: HvSearchBoxProps): JSX.Element | null;
