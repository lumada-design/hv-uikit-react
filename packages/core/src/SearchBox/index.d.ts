declare module '@hv/uikit-react-core/dist/SearchBox' {
  import React from 'react'
  
  class HvSearchBox extends React.Component<HvSearchBoxProps, any> {}

  export default HvSearchBox

  export interface SearchBoxLabel {
    inputLabel?: string
    placeholder?: string
  }

  export interface HvSearchBoxProps extends React.HTMLAttributes<HvSearchBox> {
    /**
     * The theme passed by the provider.
     */
    theme?: any

    /**
     * A Jss Object used to override or extend the styles applied to the search box.
     */
    classes?: {
      /**
       * Styles applied to searchbox root.
       */
      root?: string
    }

    /**
     * An Object containing the various text associated with the searchbox.
     *
     * - inputLabel: the label on top of the searchbox.
     * - placeholder: the placeholder value of the searchbox.
     */
    labels?: SearchBoxLabel

    /**
     * The initial value of the searchbox
     */
    value?: string

    /**
     * The function that will be executed when the searchbox changes,
     * it receives the searchbox value
     */
    onChange?: (...args: any[]) => any

    /**
     * The function that will be executed to received an array of objects that has a label and id to create list of suggestion
     */
    suggestionListCallback?: (...ags: any[]) => any

    /**
     * The function that will be executed after selecting a value in the suggestion list
     */
    suggestionSelectedCallback?: (...args: any[]) => any

    /**
     * The function that will be executed onBlur, allows checking the validation state,
     * it receives the value and the validation state (´empty´, ´filled´, ´invalid´, ´valid´).
     */
    onBlur?: (...args: any[]) => any

    /**
     * The function that will be executed onFocus, allows checking the value state,
     * it receives the value.
     */
    onFocus?: (...args: any[]) => any

    /**
     * The function that will be executed onKeyDown, allows checking the value state,
     * it receives the value.
     */
    onKeyDown?: (...args: any[]) => any

    /**
     * The function that will be executed on Enter, allows checking the value state,
     * it receives the value.
     */
    onSubmit?: (...args: any[]) => any

    /**
     * If `true` it should autofocus.
     */
    autoFocus?: boolean

    /**
     * If ´true´ the searchBox is disabled.
     */
    disabled?: boolean

    /**
     * The initial value of the searchBox.
     */
    initialValue?: string
  }
}
