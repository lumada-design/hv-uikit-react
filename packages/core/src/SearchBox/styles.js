const styles = {
  root: {
    display: "block"
  }
};

export default styles;
