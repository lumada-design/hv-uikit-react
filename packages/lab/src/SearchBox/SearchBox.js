/*
 * Copyright 2019 Hitachi Vantara Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from "react";
import PropTypes from "prop-types";

import SearchIcon from "@hv/uikit-react-icons/dist/Search.S";
import CloseIcon from "@hv/uikit-react-icons/dist/Close.S";

class HvSearchBox extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      value: props.value
    };
  }

  /**
   * Updates the state and fires callback
   */
  changeInputValue = value => {
    const { onChange, onValidate } = this.props;
    if (onValidate && !onValidate(value)) return;
    this.setState({ value });
    onChange(value);
  };

  /**
   * Handles the change of the input
   *
   * @param {Object} event - The event that triggered the action
   */
  handleInputChange = event => {
    const currentValue = event.target.value;
    this.changeInputValue(currentValue);
  };

  /**
   * Handles the clear of the input
   */
  handleInputClear = () => {
    this.changeInputValue("");
  };

  render() {
    const { placeholder, classes } = this.props;
    const { value } = this.state;

    const hasInputValue = value && value !== "";

    const iconRenderer = hasInputValue ? (
      <span onClick={this.handleInputClear}>
        <CloseIcon className={classes.iconClear} />
      </span>
    ) : (
      <SearchIcon className={classes.icon} />
    );

    return (
      <div
        className={
          hasInputValue ? classes.rootWithInput : classes.rootWithoutInput
        }
      >
        <input
          className={classes.input}
          value={value}
          placeholder={placeholder}
          type="text"
          onChange={this.handleInputChange}
        />
        {iconRenderer}
      </div>
    );
  }
}

HvSearchBox.propTypes = {
  /**
   * A Jss Object used to override or extend the styles applied to the search box.
   */
  classes: PropTypes.instanceOf(Object).isRequired,
  /**
   * The initial value of the input
   */
  value: PropTypes.string,
  /**
   * The placeholder value of the input
   */
  placeholder: PropTypes.string,
  /**
   * The function that will be executed when the input changes,
   * it receives the input value
   */
  onChange: PropTypes.func,
  /**
   * The function that use to validate input,
   */
  onValidate: PropTypes.func
};

HvSearchBox.defaultProps = {
  value: "",
  placeholder: "Search",
  onChange: () => {},
  onValidate: undefined
};

export default HvSearchBox;
