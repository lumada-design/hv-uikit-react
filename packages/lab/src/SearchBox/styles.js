/*
 * Copyright 2019 Hitachi Vantara Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

const root = theme => ({
  position: "relative",
  background: theme.hv.palette.atmosphere.atmo1,
  height: "32px",
  paddingLeft: `${theme.hv.spacing.xs}px`,
  paddingRight: `${theme.hv.spacing.md}px`
});

const icon = {
  position: "absolute",
  right: 0,
  width: "32px",
  height: "32px"
};

const styles = theme => ({
  rootWithoutInput: {
    ...root(theme),
    border: `1px solid ${theme.hv.palette.atmosphere.atmo6}`,
  },
  rootWithInput: {
    ...root(theme),
    border: `1px solid ${theme.hv.palette.accent.acce1}`,
  },

  input: {
    border: "none",
    height: "30px",
    width: "100%",
    background: "transparent",
    ...theme.hv.typography.normalText,
    "&:focus": {
      outline: "none"
    },
    "&::placeholder": {
      ...theme.hv.typography.normalText,
      color: theme.hv.typography.disabledText.color
    }
  },
  icon: {
    ...icon
  },
  iconClear: {
    ...icon,
    cursor: "pointer"
  }
});

export default styles;
