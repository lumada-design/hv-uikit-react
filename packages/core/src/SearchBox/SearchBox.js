/*  TODO: Review accessibility */

/*  eslint-disable  jsx-a11y/click-events-have-key-events */
/*  eslint-disable jsx-a11y/no-static-element-interactions */

import React, { useState } from "react";
import PropTypes from "prop-types";
import clsx from "clsx";
import partial from "lodash/partial";
import isNil from "lodash/isNil";
import { withStyles } from "@material-ui/core";
import { Search as SearchIcon } from "@hitachivantara/uikit-react-icons";
import withLabels from "../withLabels";
import { isKeypress, KeyboardCodes } from "../utils/KeyboardUtils";
import HvInput from "../Input";
import styles from "./styles";

const DEFAULT_LABELS = {
  inputLabel: "",
  placeholder: "Search",
};

/**
 *  Checks whether the user pressed Enter and executes on submit, otherwise it executes onKeyDown.
 *
 *  @param handlers - array with the callbacks to use
 *  @param event - contains the onKeyDown event
 *  @param value - the value inside the input
 */
const onKeyDownHandler = (handlers, event, value) => {
  const [onKeyDown, onSubmit] = handlers;
  if (isKeypress(event, KeyboardCodes.Enter)) {
    onSubmit?.(event, value);
  }
  onKeyDown?.(event, value);
};

/**
 *  Checks whether the user pressed Enter and executes on submit, otherwise it executes onKeyDown.
 *
 *  @param disabled - contains the onKeyDown event
 */
const changeIconColor = (disabled) => (
  <SearchIcon
    color={disabled ? "atmo7" : undefined}
    boxStyles={{
      width: "30px",
      height: "30px",
    }}
  />
);

const iconStateHandler = (value, disabled, setIcon) => {
  if (isNil(value) || value === "") {
    setIcon(changeIconColor(disabled));
  } else {
    setIcon(undefined);
  }
};

/**
 *  Checks whether the user pressed Enter and executes on submit, otherwise it executes onKeyDown.
 *
 *  @param contextValues - array with the callback to use the disabled flag.
 *  @param value - the value inside the input.
 */
const onChangeHandler = (contextValues, event, value) => {
  const [onChange, disabled, setIcon] = contextValues;
  let newValue = value;
  newValue = onChange(event, newValue);
  iconStateHandler(value, disabled, setIcon);
  return newValue;
};

/**
 *  Clears the lens icon.
 *
 *  @param contextValues - array with the callback to use the disabled flag.
 *  @param value - the value inside the input.
 */
const onFocusHandler = (contextValues, value) => {
  const [onFocus, disabled, setIcon] = contextValues;
  iconStateHandler(value, disabled, setIcon);
  onFocus?.(value);
};

/**
 *  Puts the lens icon back in place.
 *
 *  @param contextValues - array with the callback to use the disabled flag.
 *  @param value - the value inside the input.
 */
const onBlurHandler = (contextValues, value) => {
  const [onBlur, disabled, setIcon] = contextValues;
  setIcon(changeIconColor(disabled));
  onBlur?.(value);
};

const HvSearchBox = (props) => {
  const {
    classes,
    id,
    className,
    labels,
    initialValue = "",
    value,
    onChange = (event, payload) => payload,
    disabled = false,
    suggestionListCallback,
    suggestionSelectedCallback,
    onBlur,
    onFocus,
    onKeyDown,
    onSubmit,
    autoFocus = false,
    ariaLabel = "search",
    ...others
  } = props;

  const [lensIcon, setIcon] = useState(changeIconColor(disabled));

  return (
    <>
      <HvInput
        className={clsx(className, classes.root)}
        labels={labels}
        id={id}
        initialValue={initialValue}
        value={value}
        suggestionListCallback={suggestionListCallback}
        suggestionSelectedCallback={suggestionSelectedCallback}
        customFixedIcon={lensIcon}
        onChange={partial(onChangeHandler, [onChange, disabled, setIcon])}
        onBlur={partial(onBlurHandler, [onBlur, disabled, setIcon])}
        onFocus={partial(onFocusHandler, [onFocus, disabled, setIcon])}
        onKeyDown={partial(onKeyDownHandler, [onKeyDown, onSubmit])}
        autoFocus={autoFocus}
        disabled={disabled}
        showInfo={false}
        validationIconVisible={false}
        inputProps={{ "aria-label": ariaLabel }}
        {...others}
      />
    </>
  );
};

HvSearchBox.propTypes = {
  /**
   * Class names to be applied.
   */
  className: PropTypes.string,
  /**
   * A Jss Object used to override or extend the styles applied to the search box.
   */
  classes: PropTypes.shape({
    /**
     * Styles applied to searchbox root.
     */
    root: PropTypes.string,
  }).isRequired,
  /**
   * Id to be applied to the root node.
   */
  id: PropTypes.string,
  /**
   * An Object containing the various text associated with the searchbox.
   */
  labels: PropTypes.shape({
    /**
     * Label on top of the searchbox.
     */
    inputLabel: PropTypes.string,
    /**
     * Placeholder value.
     */
    placeholder: PropTypes.string,
  }),
  /**
   * The initial value of the searchbox
   */
  value: PropTypes.string,
  /**
   * The function that will be executed when the searchbox changes,
   * it receives the searchbox value
   */
  onChange: PropTypes.func,
  /**
   * The function that will be executed to received an array of objects that has a label and id to create list of suggestion
   */
  suggestionListCallback: PropTypes.func,
  /**
   * The function that will be executed after selecting a value in the suggestion list
   */
  suggestionSelectedCallback: PropTypes.func,
  /**
   * The function that will be executed onBlur, allows checking the validation state,
   * it receives the value and the validation state (´empty´, ´filled´, ´invalid´, ´valid´).
   */
  onBlur: PropTypes.func,
  /**
   * The function that will be executed onFocus, allows checking the value state,
   * it receives the value.
   */
  onFocus: PropTypes.func,
  /**
   * The function that will be executed onKeyDown, allows checking the value state,
   * it receives the value.
   */
  onKeyDown: PropTypes.func,
  /**
   * The function that will be executed on Enter, allows checking the value state,
   * it receives the value or event+value. On evocation we check for the number
   * of arguments associated with the function. Note that in the future rather
   * than accepting a value as argument, this will change to an event as argument
   */
  onSubmit: PropTypes.func,
  /**
   * If `true` it should autofocus.
   */
  autoFocus: PropTypes.bool,
  /**
   * If ´true´ the searchBox is disabled.
   */
  disabled: PropTypes.bool,
  /**
   * The initial value of the searchBox.
   */
  initialValue: PropTypes.string,
  /**
   * Arial Label.
   */
  ariaLabel: PropTypes.string,
};

export default withStyles(styles, { name: "HvSearchBox" })(withLabels(DEFAULT_LABELS)(HvSearchBox));
