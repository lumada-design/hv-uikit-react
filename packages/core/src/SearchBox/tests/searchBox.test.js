/* eslint-env jest */

import React from "react";
import { shallow, mount } from "enzyme";
import SearchBox from "../index";
import HvProvider from "../../Provider";

describe("<SearchBox />", () => {
  const onChangeMock = jest.fn();
  const onSubmitMock = jest.fn();

  it("has default props", () => {
    const wrapper = mount(
      <HvProvider>
        <SearchBox onChange={onChangeMock} />
      </HvProvider>
    );
    const props = wrapper.find(SearchBox).props();
    expect(props.value).toEqual(undefined);
    expect(props.onChange).toBe(onChangeMock);
  });

  it("renders correctly", () => {
    const wrapper = shallow(
      <HvProvider>
        <SearchBox onChange={onChangeMock} />
      </HvProvider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("renders disabled correctly", () => {
    const wrapper = shallow(
      <HvProvider>
        <SearchBox onChange={onChangeMock} disabled />
      </HvProvider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("submits on enter", () => {
    const wrapper = mount(
      <HvProvider>
        <SearchBox onChange={onChangeMock} onSubmit={onSubmitMock} />
      </HvProvider>
    );
    wrapper
      .find(SearchBox)
      .children()
      .children()
      .props()
      .onKeyDown({ keyCode: 13 }, "value");
    expect(onSubmitMock).toHaveBeenCalled();
  });

  it("renders correctly with provided search input", () => {
    const testSearchInput = "searchInput";
    const wrapper = mount(
      <HvProvider>
        <SearchBox onChange={onChangeMock} value={testSearchInput} />
      </HvProvider>
    );
    const props = wrapper.find(SearchBox).props();
    expect(props.value).toEqual(testSearchInput);
  });
});
