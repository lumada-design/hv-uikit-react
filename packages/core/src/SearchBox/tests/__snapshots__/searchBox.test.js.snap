// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBox /> renders correctly 1`] = `
<JssProvider
  generateClassName={[Function]}
>
  <HvProvider
    changeTheme={[Function]}
    router={null}
    theme={null}
    uiKitTheme="dawn"
  >
    <WithStyles(HvSearchBox)
      onChange={[MockFunction]}
    />
  </HvProvider>
</JssProvider>
`;

exports[`<SearchBox /> renders disabled correctly 1`] = `
<JssProvider
  generateClassName={[Function]}
>
  <HvProvider
    changeTheme={[Function]}
    router={null}
    theme={null}
    uiKitTheme="dawn"
  >
    <WithStyles(HvSearchBox)
      disabled={true}
      onChange={[MockFunction]}
    />
  </HvProvider>
</JssProvider>
`;
